Todo App Project (#121)
codedex-io/projects: projects/build-a-todo-app-and-setup-js-debugger-in-vscode/build-a-todo-app-and-setup-js-debugger-in-vscode.mdx

### Set up the debugger

In order for VS Code to debug your javascript code, you will have to either _launch_ your app through VS Code’s built in debugger, or _attach_ it. In this project we will go over what both of these terms mean, but it’s important to distinguish that we will be launching javascript scripts, while we will be attaching the debugger for the to-do list we will make.

First make sure you see the extension in your VS Code. If it isn't already in your Activity Bar, then you can right click your Activity Bar, and then select “Run and Debug.” You can also press Ctrl+Shift+D to see the extension.

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/activity-bar-example.png"
  description="activity-bar-example"
/>

After clicking into the extension, you should see the following options.

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/activity-bar-example-2.png"
  description="activity-bar-example-2"
/>

### Workspace Setup

> Now that your debugger is at a good state, you can refer to the project’s repo to set up your workspace. The `main` branch will contain the three files we will need for the project. You can either copy the files from the repo, or you can fork the project and start your work from `main.`

Reviewer: Not sure if I missed it, but was there a link to the project's repo provided in this file?

3. Validate the information being passed in. We will verify whether the number of people, the description, and the due date make sense. For example, it wouldn’t make sense if zero people were assigned to a task. Technically at minimum one person needs to be assigned to a task. In addition, a task can not have a blank description, title or no due date whatsoever.

### Paradigm

In this project, we will use functional programming as our strategy to make the todo list. This means a couple things:

1. Although we need to create tasks and delete tasks, we will not change or mutate the state of these to-dos as we pass them through our functions.
2. We will implement higher-order functions, which can take in other functions as a parameter. We will create a function to validate our tasks before they are rendered on the screen.

### Adding Todos

Since the project already provides the needed HTML and styling, we can immediately jump into the javascript. In this step we are going to create a function that will add todos to our todo list memory. In this case we have defined the memory as an empty array.

> Right now we are triggering the function through the typical pattern of using `.addEventListener` on the form element, but how do we convert the `event` object into the data we need? We could add a `console.log` to see what an `event` looks like, but how about we use our debugger instead? Let’s try it!

Reviewer: ```suggestion
Right now we are triggering the function through the typical pattern of using `.addEventListener()` on the form element, but how do we convert the `event` object into the data we need? We could add a `console.log()` to see what an `event` looks like, but how about we use our debugger instead? Let’s try it!
```

> Let’s add a “breakpoint” in `addTodo()` beside `event.preventDefault()`. You can do this by clicking to the left of the line number. You’ve done it correctly when you see a red dot appear.

Reviewer: ```suggestion
Let’s add a “breakpoint” in `.addTodo()` beside `event.preventDefault()`. You can do this by clicking to the left of the line number. You’ve done it correctly when you see a red dot appear.
```

Once you have the breakpoint, let’s spin up the web app in the `JavaScript Debug Terminal` by running `serve` in the terminal. You can find the `JavaScript Debug Terminal` by clicking on it in the `Run and Debug`. If the option isn’t there, you can find it by clicking on the “Launch Profile” button in your terminal. You should see a similar message blow:

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/js-debugger-terminal.png"
  description="js-debugger-terminal"
/>

Once you have the `JavaScript Debug Terminal`, and have ran `serve` in the terminal, a link will appear saying where you can access the app. Once you click that link you should see the following:

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/todo-app-form.png"
  description="todo-app-form"
/>

> Nice! Now that we have launched our app in the debugger, and our breakpoint set, let’s trigger the `addTodo` method by clicking on the “Add Todo” button. If you look back at your VS Code, you'll see that our breakpoint was triggered! As a result, we can see what some of the variables are at this point in the `addTodo` method, such as `event`.

Reviewer: ```suggestion
Nice! Now that we have launched our app in the debugger, and our breakpoint set, let’s trigger the `.addTodo()` method by clicking on the “Add Todo” button. If you look back at your VS Code, you'll see that our breakpoint was triggered! As a result, we can see what some of the variables are at this point in the `.addTodo()` method, such as `event`.
```

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/event-object-breakpoint.png"
  description="event-object"
/>

Let’s take a look at `event.target`:

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/event-target-shape.png"
  description="event-object"
/>

> It looks like `event.target` has conveniently captured all the elements in our form. That means, we could call on these elements individually and obtain their `.value` properties. Let’s try modifying our `addTodo` function like this:

Reviewer: ```suggestion
It looks like `event.target` has conveniently captured all the elements in our form. That means, we could call on these elements individually and obtain their `.value` properties. Let’s try modifying our `.addTodo()` function like this:
```

I also noticed that you use `.addTodo()` **function** here whereas you would previously use `.addTodo()` **method**. Let's just use either function or method.

```js
function addTodo(event) {
  event.preventDefault();
  const todoTitle = event.target[0].value;
  const todoNumOfPeople = event.target[1].value;
  const todoDate = event.target[2].value;
  const todoDescription = event.target[3].value;
  const todo = {
    todoTitle,
    todoNumOfPeople,
    todoDate,
    todoDescription,
  };
  const newTodos = [...todos, todo];
  return newTodos;
}
```

> You might be wondering why we are copying the `todos` array with the spread operator. In order to reduce side effects introduced directly by our functions, we need to copy the passed in `todos` array, and then return the copy. This ties into why we created an anonymous function as the listener for the `.addEventListener()` , because it then allowed us to pass in the original `todos` array.

Reviewer: ```suggestion
You might be wondering why we are copying the `todos` array with the spread operator. In order to reduce side effects introduced directly by our functions, we need to copy the passed in `todos` array, and then return the copy. This ties into why we created an anonymous function as the listener for the `.addEventListener()` function, because it then allowed us to pass in the original `todos` array.
```

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/debugger-widget.png"
  description="event-object"
/>

Now that we are adding tasks to our state, let’s look into adding them to our HTML. Let’s do this by adding two functions: `generateTodoHTM()L`and `renderTodos()`. Following the same pattern, let’s create `renderTodos()` so that it accepts `todos` and then loads those todos onto the screen.

```js
function renderTodos(todos) {
  const todoListHTML = todos.map(generateTodoHTML).join("");
  todoListElement.innerHTML = todoListHTML;
}
```

Now, for `generateTodoHTML()`, we will create a function that passes the HTML of a task item in the form of a string to `renderTodos()`.

```js

> function generateTodoHTML(todo) {
>   return `
>    <div class="card-body">
>                         <h5 class="card-title">${todo.todoTitle}</h5>
>                         <h6 class="card-subtitle mb-2 text-muted">${todo.todoDate}</h6>
>                         <h6 class="card-subtitle mb-2 text-muted">${todo.todoNumOfPeople}</h6>
>                         <p class="card-text">${todo.todoDescription}</p>
>                         <button class="btn btn-primary">Done</button>
>                         <button class="btn btn-danger">Delete</button>
>                     </div>
>     `;
> }

Reviewer: ```suggestion
function generateTodoHTML(todo) {
  return `
    <div class="card-body">
      <h5 class="card-title">${todo.todoTitle}</h5>
      <h6 class="card-subtitle mb-2 text-muted">${todo.todoDate}</h6>
      <h6 class="card-subtitle mb-2 text-muted">${todo.todoNumOfPeople}</h6>
      <p class="card-text">${todo.todoDescription}</p>
      <button class="btn btn-primary">Done</button>
      <button class="btn btn-danger">Delete</button>
    </div>
  `;
}
```

```

Finally we will want to pass `renderTools()` in the event listener, and pass in the `todos` array. Like so:

```js
document.getElementById("submitTodo").addEventListener("submit", (event) => {
  todos = addTodo(event, todos);
  renderTodos(todos);
});
```

At this point, if you add the details of a todo and then press `Add Todo` you will see the to-do render on the screen.

### Removing Todos

Excellent! Now, let us try to remove tasks from our state, while also removing them from the screen as well. There are many ways to do this, but in this case we want our users to click on the “Done” button that is rendered, and then feel the satisfaction of getting something done by seeing the task disappear on the screen.

We will do this by creating another `.addEventListener()` that will intercept click events.

```js
document.getElementById(“todo-list”).addEventListener(“click”,()=>{})
```

> Although, this event listener will intercept all clicks. In order to make sure that it only works when the “Done” button is clicked, we will add an if-statement that checks the `id` attribute of the `event` object.

Reviewer: ```suggestion
However, this event listener will intercept all clicks. In order to make sure that it only works when the “Done” button is clicked, we will add an `if`-statement that checks the `id` attribute of the `event` object.
```

```js
document.getElementById("todo-list").addEventListener("click", (event) => {
  if (event.target.id === "done") {

>     // Core logic to remove task from state and then re-render todo list

Reviewer: ```suggestion
    // Code to remove task from state and then re-render todo list
```

    errors.push("Date must be in the future");
  }
  if (todo.todoDescription.length < 10) {
    errors.push("Description must be at least 10 characters long");
  }
  return errors;
}
```

Notice that we are immediately assigning `errors` to an empty array. This is to make sure we clear the array between entries. Now, let's implement a callback function in `addTodo()` so that if the output of the callback function is an array with a length greater than zero, the task won’t be added to our state, and instead it will create an alert with the list of errors it found.

```js
function addTodo(event, todos, validateCallback) {
  event.preventDefault();
  const todoTitle = event.target[0].value;
  const todoNumOfPeople = event.target[1].value;
  const todoDate = event.target[2].value;
  const todoDescription = event.target[3].value;
  const todo = {
    todoTitle,
    todoNumOfPeople,
    todoDate,
    todoDescription,
  };
  errors = validateCallback(todo, errors);
  if (errors.length > 0) {
    alert("Uh oh! There are some errors in your form \n" + errors.join("\n"));
    return todos;
  }
  const newTodos = [...todos, todo];
  return newTodos;
}
```

> Let’s test our software! Let’s create scenarios that will target the various if statements we created in `validateTodo()`. We can do this by first filling out the form, except its title. Once we submit the form, an alert should appear saying that the title needs to have a minimum length of 5 characters.

Reviewer: ```suggestion
Let’s test our software by creating scenarios that target the various `if`-statements we created in `validateTodo()`. We can do this by first filling out the form, except its title. Once we submit the form, an alert should appear saying that the title needs to have a minimum length of 5 characters.
```

What happens if the user didn’t enter any date at all?
What happens if the user entered a date in the past?

While for the number of people assigned to the task, we need to test if we can add less than 1 person to the task, as well not assigning anyone at all.

Uh oh! Our app isn’t displaying the correct errors for scenarios 1 and 2 of our date field, and are even rendering todos with no dates and even dates in the past! Let’s add a breakpoint at `if(todo.todoDate)` and then recreate these scenarios.

In scenario one, it looks like not entering a date at all assigns an empty string to `todo.todoDate`!

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/todo-empty-string.png"
  description="event-object"
/>

While for dates in the past, it looks like the input field is assigning a string to `todo.todoDate` and then comparing it to a number (this number is the output of `Date.now()`).

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/todo-string.png"
  description="event-object"
/>

> Taking a look at `Date.now()` it looks like we need to modify our approach. `Date.now()` is a function that returns the number of seconds that have elapsed since 1 January 1970. This explains why our if-statement isn’t stopping users from picking dates in the past, because it’s essentially comparing a string to a number. Instead of storing the output of our date field directly in a variable, we need to package it in a `Date` object. That way, we can use JS’ helpful functions to convert the string to epoch time. We can do this by simply passing the value of `event.target.value` to the `Date` object before assigning it to `todoDate` in the `addTodo()` function. Your `addTodo()` function should look like the code blow

Reviewer: ```suggestion
It looks like we need to modify our approach. The `Date.now()` function returns the number of seconds that have elapsed since January 1st, 1970. This explains why our `if`-statement isn’t stopping users from picking dates in the past, because it’s essentially comparing a string to a number. Instead of storing the output of our date field directly in a variable, we need to package it in a `Date` object. That way, we can convert the string to epoch time. We can do this by simply passing the value of `event.target.value` to the `Date` object before assigning it to `todoDate` in the `addTodo()` function. Your `addTodo()` function should look like the code below:
```

    alert("Uh oh! There are some errors in your form \n" + errors.join("\n"));
    return todos;
  }
  const newTodos = [...todos, todo];
  return newTodos;
}
```

Now that `todo.todoDate` is assigned to a `Date` object, we can call on some built-in functions, like `.getTime()` which will return the epoch time of the date string we passed into the object originally. We’ll add this function to the if-statement in `validateTodo()`, and then add a breakpoint to the if-statement to test our code. In this case we will test scenario 2, entering a date in the past.

Our breakpoint should be activated. If we step over this line, we should see our app push the string “Date must be in the future” to the `errors` array. If we press play we will even see the alert box appear with the aforementioned message. Now let’s try scenario 1, which is when the user doesn’t enter a date at all. We will also move the breakpoint to `.getTime()` so that we can see the value assigned to `todo.todoDate`.

<RoundedImage
  link="https://raw.githubusercontent.com/codedex-io/projects/main/projects/build-a-todo-app-and-setup-js-debugger-in-vscode/todo-invalid-date.png"
  description="event-object"
/>

> It looks like passing an empty string to `Date()` is causing it to assign a string of “Invalid Date” to `todo.todoDate` . We will quickly resolve this by adding another if statement that checks if `todo.todoDate` equals “Invalid Date”, then push the error message “Date must be in the format of YYYY-MM-DD" to the `errors` array.

Reviewer: ```suggestion
It looks like passing an empty string to `Date()` is causing it to assign a string of “Invalid Date” to `todo.todoDate`. We will resolve this by adding another `if`-statement that checks if `todo.todoDate` is equal to “Invalid Date”, then push the error message “Date must be in the format of YYYY-MM-DD" to the `errors` array.
```

> Your ` validateTodo()` should look like this:

Reviewer: ```suggestion
Your `validateTodo()` should look like this:
```
